cmd/lotus-storage-miner/sectors.go

```diff
@@ -440,69 +444,155 @@ var sectorsExtendCmd = &cli.Command{
+					p, err := api.StateSectorPartition(ctx, maddr, si.SectorNumber, types.EmptyTSK)
+					if err != nil {
+						return xerrors.Errorf("getting sector location for sector %d: %w", si.SectorNumber, err)
+					}
+
+					if p == nil {
+						return xerrors.Errorf("sector %d not found in any partition", si.SectorNumber)
+					}
+
+					es, found := extensions[*p]
+					if !found {
+						ne := make(map[abi.ChainEpoch][]uint64)
+						ne[newExp] = []uint64{uint64(si.SectorNumber)}
+						extensions[*p] = ne
+					} else {
+						added := false
+						for exp := range es {
+							if closeEnough(exp, newExp) {
+								es[exp] = append(es[exp], uint64(si.SectorNumber))
+								added = true
+								break
+							}
+						}
+
+						if !added {
+							es[newExp] = []uint64{uint64(si.SectorNumber)}
+						}
+					}
+				}
 			}
+			p := &miner3.ExtendSectorExpirationParams{}
+			scount := 0
+			for l, exts := range extensions {
+				for newExp, numbers := range exts {
+					scount += len(numbers)
+					if scount > miner.AddressedSectorsMax || len(p.Extensions) == miner3.DeclarationsMax {
+						params = append(params, *p)
+						p = &miner3.ExtendSectorExpirationParams{}
+						scount = len(numbers)
+					}
 
-			if p == nil {
-				return xerrors.Errorf("sector %d not found in any partition", id)
+					p.Extensions = append(p.Extensions, miner3.ExpirationExtension{
+						Deadline:      l.Deadline,
+						Partition:     l.Partition,
+						Sectors:       bitfield.NewFromSet(numbers),
+						NewExpiration: newExp,
+					})
+				}
 			}
+			params = append(params, *p)
 
-			sectors[*p] = append(sectors[*p], id)
-		}
+		} else {
+			if !cctx.Args().Present() || !cctx.IsSet("new-expiration") {
+				return xerrors.Errorf("must pass at least one sector number and new expiration")
+			}
+			sectors := map[miner.SectorLocation][]uint64{}
 
-		params := &miner0.ExtendSectorExpirationParams{}
-		for l, numbers := range sectors {
+			for i, s := range cctx.Args().Slice() {
+				id, err := strconv.ParseUint(s, 10, 64)
+				if err != nil {
+					return xerrors.Errorf("could not parse sector %d: %w", i, err)
+				}
 
-			params.Extensions = append(params.Extensions, miner0.ExpirationExtension{
-				Deadline:      l.Deadline,
-				Partition:     l.Partition,
-				Sectors:       bitfield.NewFromSet(numbers),
-				NewExpiration: abi.ChainEpoch(cctx.Int64("new-expiration")),
-			})
-		}
+				p, err := api.StateSectorPartition(ctx, maddr, abi.SectorNumber(id), types.EmptyTSK)
+				if err != nil {
+					return xerrors.Errorf("getting sector location for sector %d: %w", id, err)
+				}
 
-		sp, err := actors.SerializeParams(params)
-		if err != nil {
-			return xerrors.Errorf("serializing params: %w", err)
+				if p == nil {
+					return xerrors.Errorf("sector %d not found in any partition", id)
+				}
+
+				sectors[*p] = append(sectors[*p], id)
+			}
+
+			p := &miner3.ExtendSectorExpirationParams{}
+			for l, numbers := range sectors {
```

Comment:
Can we dedup this with the above loop? It's doing the exact same thing (except now we aren't sending multiple messages).

Reply:
Yes, but eh.

Reply:
Eh == we aren't sending multiple messages here so it'll fail in some cases (although I guess we can say that that's pretty unlikely).